Add support for driver dispatch testing 2/2

ronald-cron-arm opened the pull request:
## Description
Fixes #61
Based on #587

## PR checklist
- [x] **changelog** not required because: additional testing
- [x] **framework PR** provided Mbed-TLS/mbedtls-framework#243
- [x] **mbedtls development PR** not required because: TF-PSA-Crypto only work
- [x] **mbedtls 3.6 PR**  not required because: TF-PSA-Crypto only work
- **tests**  provided: additional crypto tests

ronald-cron-arm committed:
Add TF_PSA_CRYPTO_TEST_DRIVER CMake option

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add hash acceleration component

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add paths to the `crypto_builtin_*.h` test driver headers

Add the paths to the `crypto_builtin_*.h` test driver headers
to the headers that gather the operation contexts for all
drivers.

The test-driver-specific code is guarded by
`TF_PSA_CRYPTO_TEST_LIBTESTDRIVER1`, with `libtestdriver1` as
the test driver name, using the same driver name as in Mbed TLS.

Keep the paths to the test driver headers that are still used
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
build_info.h: Add config adjustments for the test driver

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add generate_test_driver.py

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Build test driver tree

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Prefix exposed test driver identifiers

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

gilles-peskine-arm commented:
The commits are nicely split, but even so, 17 commits is on the large side. Would it make sense to split this PR just before "Add generate_test_driver.py"?

ronald-cron-arm commented:
> The commits are nicely split, but even so, 17 commits is on the large side. Would it make sense to split this PR just before "Add generate_test_driver.py"?

I think it is better to have the whole stack to better understand the why of the first commits but I can split the PR into two as you suggested. I will do that tomorrow.

ronald-cron-arm commented:
I've created #587 now with the first 10 commits. #587 should be reviewed first.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+
+import scripts_path # pylint: disable=unused-import
+from mbedtls_framework import build_tree
+from mbedtls_framework import test_driver
```
There's more code in the crypto repository than I would expect, i.e. more code that won't be shared between branches of TF-PSA-Crypto. It makes sense to me that the list of files is here, but at first glance, I'd say all of the code belongs in the framework. Can you please explain how you decided the framework/consumer split?

ronald-cron-arm replied:
My guidelines for the split between TF-PSA-Crypto and the framework:
- no knowledge of the repo tree in the framework but that a driver directory contains an `include` and `src` directory ...
- I was not really sure about the need for the "--list-vars-for-cmake" option thus I kept it and its associated code in TF-PSA-Crypto.

gilles-peskine-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -0,0 +1,45 @@
+support_universal_ctags_5_9_0() {
+    version="$(ctags --version 2>/dev/null)" || return 1
+    [[ "$version" == *"Universal Ctags 5.9.0"* ]]
+}
```
That's a very specific requirement: it has to be Universal Ctags (not Exuberant Ctags), and it can't be a newer or older version. Does it really need to be that specific?

I tried running `all.sh test_accel_hash` on Ubuntu 16.04 (with Exuberant Ctags: Universal wasn't in Ubuntu yet) plus Python 3.6, and it passed. It also passes with Exuberant Ctags 5.8 (the last official release from 2009).

How about just
```
support_ctags () {
  [[ "$(ctags --version)" = *Exuberant* ]] 2>/dev/null
}
```

That would match Exuberant Ctags since it froze in 2011, and every release of Universal Ctags so far, while excluding Emacs and BSD ctags. (The command line in `test_driver.py` doesn't work with those; it might be possible to make it work, but I'd say not worth it.)

ronald-cron-arm replied:
That's the only way I found so far to get ctags running successfully on the CI. The ctags command was falling on Ubuntu 16.04 CI even if the command is supposed to be compatible with the Exuberant version you mention above. But I can dig a bit more.

gilles-peskine-arm reviewed (commented):
Just some preliminary observations. I'm not doing an actual code review yet.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+    "tf_psa_crypto_",
+}
+
+def iter_code_files(root: Path) -> Iterable[Path]:
```
Duplicated from `test_driver.py`.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+def get_src_relpaths(builtin: Path, exclude_files: Set[str]) -> List[Path]:
+    """
+    Return the relative paths of all *.c and *.h files under `builtin`,
+    excluding those whose names match any of the patterns in `exclude_files`.
```
“file name” either refers to the base name or to the full path (not necessarily absolute), depending on who you ask. Please clarify.


```suggestion
    excluding those whose basenames match any of the patterns in `exclude_files`.
```

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+
+    out = []
+    for file in iter_code_files(builtin):
+        if not any(fnmatch(file.name, pattern) for pattern in exclude_files):
```
The same pattern appears in `test_driver.py`. It should be an auxiliary function:

```
return [path.relative_to(builtin) for path in iter_code_files_with_excludes(builtin, exclude_files)]
```

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+    out.sort()
+    return out
+
+def get_dst_relpaths(src_relpaths: List[Path], driver: str) -> List[Path]:
```
This duplicates some of the logic in `test_driver.py` and should be unified.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+                                          for directory in directories)
+    identifiers = set()
+    for file in files:
+        identifiers.update(test_driver.run_ctags(file))
```
Why not run `ctags` on the list of files in a single batch?

ronald-cron-arm replied:
Because the command would contain a long list of files.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+    Main function of this program
+    """
+    parser = test_driver.get_parsearg_base()
+    parser.add_argument('--list-vars-for-cmake', nargs="?", const="__AUTO__", metavar="FILE",
```
Please document that this is done _instead_ of creating a test driver tree, not in addition.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+
+    dst_dir = Path(args.dst_dir)
+    if not dst_dir.is_dir():
+        raise RuntimeError(f"Do not find expected {args.dst_dir} directory.")
```
Why doesn't the script create the destination directory? That's what I would have expected.
```
git checkout …
tests/scripts/generate_test_driver.py drivers/testdriver1
tests/scripts/generate_test_driver.py drivers/testdriver2
```

ronald-cron-arm replied:
This is more convenient for the integration with CMake. At least with the way I've done it.
During the configuration phase, CMake creates the libtestdriver1 directory in the build tree, adds a CMakeLists.txt into it and then adds it as a subdirectory of drivers: essentially add_subdirectory(libtestdriver1), like we have add_subdirectory(builtin).
The CMakeLists.txt in `libtestdriver1` is roughly just the `builtin/CMakeLists.txt` one (as it is inside `libtestdriver1/`) plus the code to generate the test driver code during the build  phase.
Thus `generate_test_driver.py`  is called when `libtestdriver1` has already been created and contains its CMakeLists.txt file.

gilles-peskine-arm replied:
As of 2b70a5e35e5b5b92aa24295cfbaf166db61160e8, the script works fine whether the destination directory exists or not, except for this explicit check. So please remove it.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+
+        with open(dst_dir / fname, "w") as f:
+            f.write(f"set({args.driver}_input_files " + \
+                     " ".join(str(path) for path in src_relpaths) + ")\n\n")
```
To facilitate debugging, it would be nicer to have one file per line.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+                    for path in get_dst_relpaths(src_relpaths, args.driver)) + ")\n\n")
+            f.write(f"set({args.driver}_src_files " + \
+                    " ".join(str(path) \
+                    for path in src_relpaths if path.suffix == ".c") + ")")
```
Missing final newline.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+                    for path in src_relpaths if path.suffix == ".c") + ")")
+        return
+
+    #Step 1: Build the test driver tree from `drivers/builtin`
```
Both here and in `test_driver.py`, “build” is misleading: in this context, it tends to refer to invoking a C compiler. “Build tree” is even worse because as a noun, it means the directories where non-source files are placed, as opposed to the source tree. We tend to use “generate” for this kind of step that happens before compilation, but it is admittedly vague and already overused. I'd still prefer “generate” to “build”, or better, some different verb. Putting files in a directory can be “populating” it, but that strongly presumes a preexisting target directory. We could use “construct”, which is a synonym of “build” in everyday English but does not have a commonly used meaning in this context.

ronald-cron-arm replied:
We could use "populate" then as we have a preexisting target directory.

ronald-cron-arm replied:
I've changed to `create_test_driver_tree`.

gilles-peskine-arm commented on tests/scripts/generate_test_driver.py:
```diff
@@ -0,0 +1,181 @@
+        if identifier.startswith("PSA_WANT_"):
+            identifiers.add(identifier.replace("PSA_WANT_", "MBEDTLS_PSA_ACCEL_", 1))
+
+    generator.prefix_identifiers(identifiers)
```
Having reviewed `generate_test_driver.py`, I now think that all the code should be in the shared module `test_driver.py`, except:

* Branch-specific data: `EXCLUDE_FILES`, maybe `IDENTIFIER_PREFIXES`.
* The ad hoc fiddling of `identifiers` after adding the external identifiers and before calling `prefix_identifers`.
* Calling `main`.

I think this would be best handled by putting all the logic in `TestDriverGenerator`, and having a method to adjust the identifiers which does nothing by default and is overridden in a derived class defined in the branch-specific script.

ronald-cron-arm replied:
I have something along these lines now.

gilles-peskine-arm reviewed (changes requested):
I have reviewed `tests/scripts/generate_test_driver.py`, but not the rest of this PR, at c1bad1c6232855d4b37884d40b3db727321059df. This is a companion to my [review of the framework part](https://github.com/Mbed-TLS/mbedtls-framework/pull/243#pullrequestreview-3536066960).

ronald-cron-arm committed:
Update framework - first version of test_driver.py

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Remove support for list-vars-for-cmake

Move support for list-vars-for-cmake to test_driver.py
in the framework.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Define subclass of TestDriverGenerator

Define the subclass TFPSACryptoTestDriverGenerator
of TestDriverGenerator and overwrite the
`get_identifiers_to_prefix` method. The build tree is then
completely created just by calling `build_tree`.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Adjust to build_tree renaming

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm commented:
@minosgalanakis @gilles-peskine-arm , as discussed offline, currently this PR is mostly part of the validation of the framework PR. Compared to the previous version I've rebased on top of the current head of https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/587 (https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/587/commits/a548d3fadcba59f35ba4e4ed068874346c207349) and then added commits at the top to adjust to the changes in test_driver.py in the framework PR:
```
2b70a5e35 Adapt to systematic file basename prefixing
902bfbfbc generate_test_driver.py: Adjust to build_tree renaming
ae7c2c4f5 generate_test_driver.py: Define subclass of TestDriverGenerator
1f62f9b30 generate_test_driver.py: Remove support for list-vars-for-cmake
```
If you checkout one of these commits and update the associated framework commit, the CI component `test_accel_hash` passes successfully.

minosgalanakis commented on CMakeLists.txt:
```diff
@@ -391,6 +406,9 @@ if(ENABLE_TESTING OR ENABLE_PROGRAMS)
         add_custom_target(tf_psa_crypto_test_keys_header
             DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/tests/include/test/test_keys.h)
         add_dependencies(tf_psa_crypto_test tf_psa_crypto_test_keys_header)
+        if(TF_PSA_CRYPTO_TEST_DRIVER)
```
Design question. Is there any feasible use case to disable testing and programs, and still need the TF_PSA_CRYPTO_TEST_DRIVER? if so that logic should be revisitted.

ronald-cron-arm replied:
Programs yes but not testing. TF_PSA_CRYPTO_TEST_DRIVER is part of the testing.

minosgalanakis commented on tests/scripts/libtestdriver1.cmake:
```diff
@@ -0,0 +1,28 @@
+    COMMAND
+        ${TF_PSA_CRYPTO_PYTHON_EXECUTABLE}
+            ./tests/scripts/generate_test_driver.py ${CMAKE_CURRENT_BINARY_DIR}
+    DEPENDS ${PROJECT_SOURCE_DIR}/tests/scripts/generate_test_driver.py
```
Why this depedency on the script itself? I understand on why that is needed for development, but should we also regenerate the files when the input header have been touched?

Something like:

~~~~~~
add_custom_command(
    OUTPUT ${libtestdriver1_files}
    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
    COMMAND
        ${TF_PSA_CRYPTO_PYTHON_EXECUTABLE}
            ./tests/scripts/generate_test_driver.py ${CMAKE_CURRENT_BINARY_DIR}
    DEPENDS
        ${PROJECT_SOURCE_DIR}/tests/scripts/generate_test_driver.py
       ${libtestdriver1_input_files}
~~~~~~

I think we can access the output of 'generate_test_driver.py --list-vars-for-cmake' at this stage?

minosgalanakis reviewed (commented):
Looks good overall, and no big objections to the design. Since it is in the bottom of the PR depedency chain, I am only commenting at this stage

ronald-cron-arm committed:
Update framework - latest version of test_driver.py

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Prefix MBEDTLS_PK_PARSE_EC_EXTENDED/COMPRESSED

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
generate_test_driver.py: Remove dst_dir existence constraint

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm commented:
@gilles-peskine-arm @minosgalanakis I've rebased on top of development after the merge #587. I've kept the history of the work on generate_test_driver.py alongside the work on test_driver.py in the framework t the bottom of the commit stack. I believe as part of the review of https://github.com/Mbed-TLS/mbedtls-framework/pull/243 you have already reviewed up to the commit "Update framework - latest version of test_driver.py". I also believe I have addressed all the comments I received so far. The CI is green this the PR is ready for a full review now.

gilles-peskine-arm commented on core/CMakeLists.txt:
```diff
@@ -171,6 +171,11 @@ foreach(target IN LISTS target_libraries)
+    if(TF_PSA_CRYPTO_TEST_DRIVER)
+        target_include_directories(${target}
+            PRIVATE ${TF_PSA_CRYPTO_FRAMEWORK_DIR}/tests/include
+                    ${TF_PSA_CRYPTO_DIR}/tests/include/test)
```
Why are we putting `tests/include/test` and not `tests/include` on the include path?

ronald-cron-arm replied:
This is the specific path for "mbedtls/build_info.h" that is included by the framework test driver headers, like in `framework/tests/include/test/drivers/aead.h`.

gilles-peskine-arm replied:
Ah, ok. Please explain this in a comment.

gilles-peskine-arm commented on drivers/CMakeLists.txt:
```diff
@@ -26,6 +25,22 @@ if(CMAKE_COMPILER_IS_MSVC)
 
+if(TF_PSA_CRYPTO_TEST_DRIVER)
+    # Use `libtestdriver1` as the driver name to align with the name used in
+    # Mbed TLS. This allows to reuse most of the test driver code as it is.
```
I think that's in reference to Mbed TLS 3.6, right? In context I would understand it as referring to the version of Mbed TLS that matches the version of TF-PSA-Crypto that I'm reading, but I don't see why the content of that version of Mbed TLS would matter here.

ronald-cron-arm replied:
I've tried and improved the comment.

gilles-peskine-arm commented on drivers/CMakeLists.txt:
```diff
@@ -26,6 +25,22 @@ if(CMAKE_COMPILER_IS_MSVC)
+        COPYONLY
+    )
+    add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/libtestdriver1
+                     ${CMAKE_CURRENT_BINARY_DIR}/libtestdriver1)
```
Please explain what's happening here, and how `tests/scripts/libtestdriver1.cmake` gets used. Even the fact that these are the same question isn't fully obvious.

As a CMake newbie, I think what's going on is:

1. We copy `tests/scripts/libtestdriver1.cmake` somewhere in the build tree, and we choose to give it the basename `CMakeLists.txt`.
2. We tell CMake to treat the location of this newly created `CMakeLists.txt` file as a source directory, so it reads and executes that `CMakeLists.txt`.
3. The fact that `${CMAKE_CURRENT_BINARY_DIR}/libtestdriver1` is both the source tree and the build tree for that part of the build is nothing special: it's what happens in any in-tree build, and CMake lets you choose the source/build tree correspondence independently in each subdirectory.

Please let me know if that's wrong or incomplete.

What's the benefit of this approach, compared to having `drivers/libtestdriver1/CMakeLists.txt` in the source tree in version control?

ronald-cron-arm replied:
I've added a comment. 
Your three points above are correct.
> What's the benefit of this approach, compared to having drivers/libtestdriver1/CMakeLists.txt in the source tree in version control?

This way we avoid having a test-specific directory in the library tree. We also plan to generate multiple test drivers, and this approach seems more flexible for that.

gilles-peskine-arm commented on tests/scripts/libtestdriver1.cmake:
```diff
@@ -0,0 +1,34 @@
+    COMMENT "Generating test driver libtestdriver1 tree"
+)
+
+add_custom_target(tf_psa_crypto_test_driver_src
```
So it's `make tf_psa_crypto_test_driver_src` to generate the sources but `make libtestdriver1` to compile them? Can we make that consistent?

ronald-cron-arm replied:
Changed to `libtestdriver1_generation` and added a TF_PSA_CRYPTO_TEST_DRIVER_GENERATION_TARGETS list in the model of TF_PSA_CRYPTO_DRIVER_TARGETS to avoid hard-coding `libtestdriver1` in test and drivers build scripts.

gilles-peskine-arm commented on tests/configs/user-config-test-driver-extension.h:
```diff
@@ -0,0 +1,426 @@
+
+#if defined(PSA_WANT_DH_RFC7919_2048) && \
+    !defined(MBEDTLS_PSA_ACCEL_DH_RFC7919_2048)
+#define LIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_DH_RFC7919_2048
```
For some reason the `DH_xxx` definitions have an empty expansion instead of 1 like all the others.

gilles-peskine-arm commented on tests/configs/user-config-test-driver-extension.h:
```diff
@@ -0,0 +1,426 @@
+/**
```
This file would be trivial to generate in terms of its content, and it's repetitive and error prone. With the overhaul of the libtestdriver1 generation, how hard would it be to fit its generation into the build scripts?

ronald-cron-arm replied:
We could generate it in the build tree in `drivers/libtestdriver1/include` or a sub-directory of it and this should make it. This would mean a framework PR to add the generation logic to test_driver.py. Thus I would prefer to do this later.

gilles-peskine-arm replied:
Sure, I didn't mean to do it now, sorry for being unclear. Filed as https://github.com/Mbed-TLS/mbedtls-framework/issues/255

gilles-peskine-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -0,0 +1,37 @@
+    not grep mbedtls_sha1 ${CMAKE_BUILTIN_BUILD_DIR}/sha1.c.o
+    not grep mbedtls_sha256 ${CMAKE_BUILTIN_BUILD_DIR}/sha256.c.o
+    not grep mbedtls_sha512 ${CMAKE_BUILTIN_BUILD_DIR}/sha512.c.o
+    not grep mbedtls_ripemd160 ${CMAKE_BUILTIN_BUILD_DIR}/ripemd160.c.o
```
Just noting: This should have SHA3, except that it can't due to https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/542

gilles-peskine-arm reviewed (changes requested):
I think the code is correct, but there are a few places that I'm not sure I understand and I want at least one comment.

minosgalanakis reviewed (dismissed):
LGTM.

tests/scripts/generate_test_driver.py is now  neat and explicit and easy to understand and maintain.

The build system changes are consistent with the agreed design. Thank you

p.s It still will require a rebase for the framework conflict

ronald-cron-arm committed:
cmake: Add/Improve comments

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
cmake: drivers: Introduce TF_PSA_CRYPTO_TEST_DRIVER_GENERATION_TARGETS

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
user-config-test-driver-extension.h: Fix DH_xxx definitions

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
cmake: Do not use TF_PSA_CRYPTO_DIR

Do not use TF_PSA_CRYPTO_DIR, use PROJECT_SOURCE_DIR
instead to align with the changes in
https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/531.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Update framework to the merge of #243

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm commented:
@gilles-peskine-arm @minosgalanakis I believe I've addressed all the comments. I've added five commits and had to rebase because of a framework conflict. The head before the rebase was https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/586/commits/bd41097f535607cefadec77dbca9d557c2240aec. The rebase can be checked with:
```
git range-diff HEAD~19..HEAD bd41097f535607cefadec77dbca9d557c2240aec~19..bd41097f535607cefadec77dbca9d557c2240aec
```
Only difference: the initial framework commit in the first commit that updates the framework pointer.
The CI is green, this is ready for re-review.

gilles-peskine-arm commented on tests/scripts/libtestdriver1.cmake:
```diff
@@ -0,0 +1,34 @@
+include(${CMAKE_CURRENT_BINARY_DIR}/libtestdriver1-list-vars.cmake)
+
+set(input_files "")
+foreach(relpath IN LISTS libtestdriver1_input_files)
```
I can't figure out where `libtestdriver1_input_files` is defined.

ronald-cron-arm replied:
In libtestdriver1-list-vars.cmake, I am adding a comment.

gilles-peskine-arm reviewed (changes requested):
LGTM except I'd like a couple more comments.

ronald-cron-arm committed:
Add comments

Add comments about libtestdriver1-list-vars.cmake
variables and "tests/include/test" include path.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

gilles-peskine-arm commented on tests/scripts/libtestdriver1.cmake:
```diff
@@ -9,6 +9,10 @@ if(result)
+# - libtestdriver1_input_files: the list of generate_test_driver.py input files
+# - libtestdriver1_files: the list of generate_test_driver.py output files
+# - libtestdriver1_src_files: the list of generate_test_driver.py output C files
 include(${CMAKE_CURRENT_BINARY_DIR}/libtestdriver1-list-vars.cmake)
```
`libtestdriver1-list-vars.cmake` is created by the `execute_process` just above. It's not obvious because the file name doesn't appear in the command. It is a reasonable expectation given the sequence of instructions in `libtestdriver1.cmake`, but still, I would prefer to have a comment to explain this.

Come to think of it, if `generate_test_driver.py` changes, will CMake re-generate `libtestdriver1-list-vars.cmake`?

ronald-cron-arm replied:
> `libtestdriver1-list-vars.cmake` is created by the `execute_process` just above. It's not obvious because the file name doesn't appear in the command. It is a reasonable expectation given the sequence of instructions in `libtestdriver1.cmake`, but still, I would prefer to have a comment to explain this.

I've added the file name in the command.  

> Come to think of it, if `generate_test_driver.py` changes, will CMake re-generate `libtestdriver1-list-vars.cmake`?

Not sure we need to support that but as the libtestdriver1 tree creation depends on `generate_test_driver.py`, I've updated the code to re-trigger the configuration phase (and thus the generation of `libtestdriver1-list-vars.cmake`) if `generate_test_driver.py` is modified.

gilles-peskine-arm commented on core/CMakeLists.txt:
```diff
@@ -172,6 +172,8 @@ foreach(target IN LISTS target_libraries)
     endforeach()
     if(TF_PSA_CRYPTO_TEST_DRIVER)
+        # See the note on include directories for tf_psa_crypto_test target
+        # regarding "tests/include/test".
```
Where is that note?

Ah, found it in the root `CMakeLists.txt`. Please add this information in the comment. Also, for the CMake non-experts, it would help to be more explicit that “note on include directories” means “note above the `target_include_directories(tf_psa_crypto_test` call” (`tf_psa_crypto_test` occupies about 50 lines).

ronald-cron-arm replied:
I've just duplicated the note as the pointer to the note was getting as long as the note itself.

gilles-peskine-arm reviewed (changes requested):
Sorry, still more comments please.

ronald-cron-arm committed:
libtestdriver1.cmake: Set cmake list vars file name explicitely

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
libtestdriver1.cmake: Regenerate cmake list vars file

Regenerate cmake list vars file if
generated_test_driver.py is modified.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
cmake: Improve comment

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

gilles-peskine-arm reviewed (approved):
Thanks, LGTM

ronald-cron-arm commented:
@minosgalanakis please have a look to the last version, thanks.

minosgalanakis reviewed (approved):
LGTM

Cheked the [range after ](https://github.com/Mbed-TLS/TF-PSA-Crypto/compare/128ac9724773122e8002f7d340aa0be5a404cccd..26c5b39f12ba3a6f0c25c54b08c327c09ca7e8cb) and verified the rebase locally using  `git range-diff HEAD~19..HEAD bd41097f535607cefadec77dbca9d557c2240aec~19..bd41097f535607cefadec77dbca9d557c2240aec`
